frame/offences/src/lib.rs

```diff
@@ -103,6 +144,37 @@ pub mod pallet {
 	}
 }
 
+impl<T: Config> Pallet<T> {
+	/// This is called at the start of each new session. Hence, only one session becomes
+	/// obsolete in this call, whose data gets cleared up here.
+	///
+	/// For example, if the `current_session_index` is 10 and `MaxSessionReportAge` is 6,
+	/// this clears all reports for `obsolete_session_index` 4.
+	fn clear_obsolete_reports(current_session_index: SessionIndex) {
+		if current_session_index <= T::MaxSessionReportAge::get() {
+			return
+		}
+
+		let obsolete_session_index = current_session_index - T::MaxSessionReportAge::get();
+
+		let session_reports = SessionReports::<T>::take(obsolete_session_index);
+		let session_reports =
+			Vec::<(Kind, OpaqueTimeSlot, ReportIdOf<T>)>::decode(&mut &session_reports[..])
+				.unwrap_or_default();
```

Comment:
and why `unwrap_or_default`? if we go to default, there must have been a bug, no? then it can be a `defensive_unwrap_or_default`, or an early return.

Reply:
It will go to default when the vector is empty?

Reply:
No. If it is empty it just returns `Ok(vec![])`, an error here means that the storage is corrupted.  

DQ but why is this double-encode anyway? Why not store `Vec<StorageReportOf<T>>`?